backends/p4tools/modules/testgen/targets/bmv2/map_direct_externs.cpp

```diff
@@ -1,46 +1,68 @@
 #include "backends/p4tools/modules/testgen/targets/bmv2/map_direct_externs.h"
 
+#include <cstdio>
+#include <optional>
+
+#include "ir/ir.h"
+#include "lib/error.h"
+
 namespace P4Tools::P4Testgen::Bmv2 {
 
 bool MapDirectExterns::preorder(const IR::Declaration_Instance *declInstance) {
     declaredExterns[declInstance->name] = declInstance;
     return true;
 }
 
-bool MapDirectExterns::preorder(const IR::P4Table *table) {
-    // Try to either get counters or meters from the table.
-    const auto *impl = table->properties->getProperty("meters");
-    if (impl == nullptr) {
-        impl = table->properties->getProperty("counters");
+std::optional<const IR::Declaration_Instance *> MapDirectExterns::getExternFromTableImplementation(
+    const IR::Property *tableImplementation) {
+    const auto *selectorExpr = tableImplementation->value->to<IR::ExpressionValue>();
```

Comment:
nit: Why move from `checkedTo` to just `to`? Isn't that less safe?

Reply:
Maybe this is what the nullptr case below covers?

Reply:
It's an exception vs error resolution thing. The exception will throw once and then terminate entirely. With the error we can continue execution and potentially emit other errors in this particular pass. Compilation will still stop once this compiler pass has completed.